datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,969 @@
+        self.map.len()
+    }
+
+    unsafe fn update_heap_idx(&mut self, mapper: &[(usize, usize)]) {
+        self.map.update_heap_idx(mapper);
+    }
+
+    unsafe fn heap_idx_at(&self, map_idx: usize) -> usize {
+        self.map.heap_idx_at(map_idx)
+    }
+
+    fn drain(&mut self) -> (ArrayRef, Vec<usize>) {
+        let (ids, heap_idxs) = self.map.drain();
+        let mut builder: PrimitiveBuilder<VAL> = PrimitiveArray::builder(ids.len());
+        for id in ids.into_iter() {
+            match id {
+                None => builder.append_null(),
+                Some(id) => builder.append_value(id),
+            }
+        }
+        let ids = Arc::new(builder.finish());
+        (ids, heap_idxs)
+    }
+
+    unsafe fn find_or_insert(
+        &mut self,
+        row_idx: usize,
+        replace_idx: usize,
+        mapper: &mut Vec<(usize, usize)>,
+    ) -> (usize, bool) {
+        let ids = self.owned.as_primitive::<VAL>();
+        let id: Option<VAL::Native> = if ids.is_null(row_idx) {
+            None
+        } else {
+            Some(ids.value(row_idx))
+        };
+
+        let hash: u64 = id.hash(&self.rnd);
+        if let Some(map_idx) = self.map.find(hash, |mi| id == *mi) {
+            return (map_idx, false);
+        }
+
+        // we're full and this is a better value, so remove the worst
+        let heap_idx = self.map.remove_if_full(replace_idx);
+
+        // add the new group
+        let map_idx = self.map.insert(hash, id, heap_idx, mapper);
+        (map_idx, true)
+    }
+}
+
+pub trait ValueType: Comparable + Clone + Debug {}
+
+impl<T> ValueType for T where T: Comparable + Clone + Debug {}
+
+pub trait KeyType: Clone + Comparable + Debug {}
+
+impl<T> KeyType for T where T: Clone + Comparable + Debug {}
+
+impl GroupedTopKAggregateStream {
+    fn intern(&mut self, ids: ArrayRef, vals: ArrayRef) -> Result<()> {
+        let mut mapper = Vec::with_capacity(self.limit);
+        let len = ids.len();
+        self.map.set_batch(ids);
+        self.heap.set_batch(vals.clone());
+
+        let has_nulls = vals.null_count() > 0;
+        for row_idx in 0..len {
+            if has_nulls && vals.is_null(row_idx) {
+                continue;
+            }
+            self.insert(row_idx, &mut mapper)?;
+        }
+        Ok(())
+    }
+
+    pub fn insert(
+        &mut self,
+        row_idx: usize,
+        map: &mut Vec<(usize, usize)>,
+    ) -> Result<()> {
+        assert!(self.map.len() <= self.limit, "Overflow");
+
+        // if we're full, and the new val is worse than all our values, just bail
+        if self.heap.is_worse(row_idx) {
+            return Ok(());
+        }
+
+        // handle new groups we haven't seen yet
+        map.clear();
+        let replace_idx = self.heap.worst_map_idx();
+        let (map_idx, did_insert) =
+            unsafe { self.map.find_or_insert(row_idx, replace_idx, map) };
+        if did_insert {
+            self.heap.renumber(map);
```

Comment:
As `HeapItem`s get `swap()`ed or `HashTable`s grow, we need to track the mapping of old indexes to new and update the other struct.